### PR TITLE
Update Iterative Tasking Infrastructure

### PR DESCRIPTION
**Pull request by pdmullen**
## PR Summary

I am submitting this PR on behalf of @jdolence who made these changes in `lroberts36/merge-sparse-with-jdolence-sparse` some time ago.  In practice, this PR slightly modifies the task list infrastructure to update iterative tasking, in particular with regional dependencies.  This PR fixes various failure modes exhibited in downstream codes whereby (1) iterative tasks failed to iterate, (2) hangs were experienced, or (3) unexpected results were obtained with MPI both with and without regional dependencies.  

## PR Checklist

- [x] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Review by Yurlungur — APPROVED** [review]
Thanks for porting this in. Sorry I missed it on the first batch of updates.

File: src/tasks/task_list.hpp [hunk 1]
```diff
@@ -44,17 +45,18 @@ TaskID AddTaskHelper(TaskList *, Task);
...
   // NOTE: we must capture the object pointer
-  template <class T, class U, class... Args>
-  TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args...), U *obj,
-                 Args &&...args) {
+  template <class T, class U, class... Args1, class... Args2>
+  TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args1...), U *obj,
+                 Args2 &&...args) {
     return this->AddTask_(TaskType::iterative, 1, dep, [=]() mutable -> TaskStatus {
-      return (obj->*func)(std::forward<Args>(args)...);
+      return (obj->*func)(std::forward<Args2>(args)...);
```
Comment (lroberts36) [inline src/tasks/task_list.hpp:59]: I don't immediately see why we need two separate variadic template parameters here. I am mostly just curious what issue the single list was causing. Maybe it was something to do with default arguments?

File: src/tasks/task_list.hpp [hunk 2]
```diff
@@ -393,7 +466,12 @@ class TaskRegion {
   }
 
  private:
-  bool HasRun(const int reg_id) {
+  void AddDepdencies(const std::string &label, const int list_id, const TaskID &tid) {
```
Comment (lroberts36) [inline src/tasks/task_list.hpp]:
```suggestion
  void AddDependencies(const std::string &label, const int list_id, const TaskID &tid) {
```

**Review by lroberts36 — APPROVED** [review]
LGTM, thanks for porting this. I did not go through the logic in detail, but that is probably not as useful as the test that it actually works in Riot.

File: src/tasks/task_list.hpp (hunk 1 above)
Comment (jdolence) [reply src/tasks/task_list.hpp:59]: Unfortunately I can't remember _exactly_ what situation led to this change, but I can give you the basic idea.  What happened was that I wrote a task with some set of reasonable arguments (I thought) but the compiler yelled at me about type mismatches when I actually tried to add it to a task list.  Don't hold me to this, but I think it was a `const` vs non-`const` issue.  I convinced myself at the time that there was no reason we should be preventing people from writing tasks the way I had written the problematic one, and this transformation generalized the interface to allow it.

**Comment by pdmullen** [comment]
Extended regression is failing upon enabling auto-merge. Is this definitely pointing to an issue with the changes in this PR? Or could it be unrelated? 

Asking before I jump down the rabbit hole.

**Comment by Yurlungur** [comment]
> Extended regression is failing upon enabling auto-merge. Is this definitely pointing to an issue with the changes in this PR? Or could it be unrelated?

Given that automerge eventually let it go, sounds like the issue was just the test machine overloaded.

**Pull request merged**